components/style/values/specified/length.rs

```diff
@@ -222,14 +222,96 @@ impl CharacterWidth {
     }
 }
 
+const ABSOLUTE_LENGTH_MAX: f64 = (1 << 30) as f64;
+const ABSOLUTE_LENGTH_MIN: f64 = - (1 << 30) as f64;
+
+macro_rules! impl_from_au {
+    ($ty: ty, $multiplier: expr) => {
+        impl From<$ty> for Au {
+            fn from(length: $ty) -> Au {
+                Au(
+                    ((length.0 as f64) * ($multiplier as f64))
+                    .max(ABSOLUTE_LENGTH_MIN)
+                    .min(ABSOLUTE_LENGTH_MAX)
+                    .round() as i32
+                )
+            }
+        }
+    }
+}
+
+/// A wrapper representing a floating-point pixel (px) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Px(pub f32);
+
+impl_from_au!(Px, AU_PER_PX);
+
+/// A wrapper representing a floating-point inches (in) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct In(pub f32);
+
+impl_from_au!(In, AU_PER_IN);
+
+/// A wrapper representing a floating-point centimeter (cm) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Cm(pub f32);
+
+impl_from_au!(Cm, AU_PER_CM);
+
+/// A wrapper representing a floating-point millimeter (mm) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Mm(pub f32);
+
+impl_from_au!(Mm, AU_PER_MM);
+
+/// A wrapper representing a floating-point quarter-millimeter (q) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Q(pub f32);
+
+impl_from_au!(Q, AU_PER_Q);
+
+/// A wrapper representing a floating-point point (pt) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Pt(pub f32);
+
+impl_from_au!(Pt, AU_PER_PT);
+
+/// A wrapper representing a floating-point pica (pc) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Pc(pub f32);
+
+impl_from_au!(Pc, AU_PER_PC);
+
 /// A `<length>` without taking `calc` expressions into account
 ///
 /// https://drafts.csswg.org/css-values/#lengths
 #[derive(Clone, PartialEq, Copy, Debug)]
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 pub enum NoCalcLength {
-    /// An absolute length: https://drafts.csswg.org/css-values/#absolute-length
-    Absolute(Au),  // application units
+    /// An absolute length in application units
+    /// See https://drafts.csswg.org/css-values/#absolute-length
+    Au(Au),  // application units
```

Comment:
Alternatively we could also give up and inline `FontRelativeLength` here, but unless we have a good reason to, I'd prefer not.